lib/AST/ASTImporter.cpp

```diff
@@ -931,8 +931,30 @@ bool ASTNodeImporter::ImportDeclParts(NamedDecl *D, DeclContext *&DC,
                                       DeclarationName &Name, 
                                       NamedDecl *&ToD,
                                       SourceLocation &Loc) {
+  // Check if RecordDecl is in FunctionDecl parameters to avoid infinite loop.
+  // example: int struct_in_proto(struct data_t{int a;int b;} *d);
+  DeclContext *OrigDC = D->getDeclContext();
+  FunctionDecl *FunDecl;
+  if (isa<RecordDecl>(D) && (FunDecl = dyn_cast<FunctionDecl>(OrigDC))) {
+    bool DoNotImport = !FunDecl->hasBody();
+    if (!DoNotImport) {
+      SourceRange RecR = D->getSourceRange();
+      SourceRange BodyR = FunDecl->getBody()->getSourceRange();
+      // If RecordDecl is not in Body (it is a param), we bail out.
+      DoNotImport = RecR.isValid() && BodyR.isValid() &&
+                    (RecR.getBegin() < BodyR.getBegin() ||
+                     BodyR.getEnd() < RecR.getEnd());
+    }
+    if (DoNotImport) {
```

Comment:
Why not a simple else?

Reply:
Because we can also change our mind and change DoNotImport in the then case if it turns out that we are out of range.